src/RemoteRequest/CachedResponse.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * Class CachedResponse.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\RemoteRequest;
+
+use DateTimeImmutable;
+use DateTimeInterface;
+
+/**
+ * Serializable object that represents a cached response together with its expiry time.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class CachedResponse {
+
+	/**
+	 * Cached body.
+	 *
+	 * @var string
+	 */
+	private $body;
+
+	/**
+	 * Cached headers.
+	 *
+	 * @var array
+	 */
+	private $headers;
+
+	/**
+	 * Cached status code.
+	 *
+	 * @var int
+	 */
+	private $status_code;
+
+	/**
+	 * Expiry time of the cached value.
+	 *
+	 * @var DateTimeInterface
+	 */
+	private $expiry;
+
+	/**
+	 * Instantiate a CachedResponse object.
+	 *
+	 * @param string            $body         Cached body.
+	 * @param array             $headers      Cached headers.
+	 * @param int               $status_code  Cached status code.
+	 * @param DateTimeInterface $expiry       Expiry of the cached value.
+	 */
+	public function __construct( $body, array $headers, $status_code, DateTimeInterface $expiry ) {
+		$this->body        = $body;
+		$this->headers     = $headers;
+		$this->status_code = $status_code;
+		$this->expiry      = $expiry;
```

Comment:
We should ensure somewhere that the types we claim to return are correct (we don't want to simulate WordPress too much).
For example, the `$body` could be a stream object instead of a string.
```suggestion
		$this->body        = (string) $body;
		$this->headers     = (array) $headers;
		$this->status_code = (int) $status_code;
		$this->expiry      = $expiry;
```